notebooks/Mitigating-Disparities-in-Ranking-from-Binary-Data.ipynb

```diff
@@ -0,0 +1,592 @@
+    "    \n",
+    "    def predict(self, X):\n",
+    "        # use predict_proba to get real values instead of 0/1, select only prob for 1\n",
+    "        scores = self.logistic_regression_estimator.predict_proba(X)[:,1]\n",
+    "        return scores\n",
+    "\n",
+    "\n",
+    "balanced_index_pass0 = y_train[y_train==0].index \n",
+    "balanced_index_pass1 = y_train[y_train==1].sample(n=balanced_index_pass0.size, random_state=0).index\n",
+    "balanced_index = balanced_index_pass0.union(balanced_index_pass1)\n",
+    "\n",
+    "pp_estimator = ThresholdOptimizer(\n",
+    "    unconstrained_predictor=LogisticRegressionAsRegression(unmitigated_predictor),\n",
+    "    constraints=\"demographic_parity\")\n",
+    "\n",
+    "pp_estimator.fit(X_train.iloc[balanced_index,:], y_train.iloc[balanced_index],\n",
+    "                 sensitive_features=A_train.iloc[balanced_index])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We next evaluate AUC (balanced accuracy) and demographic disparity (selection rates) of the black and white students on the test data; note that we use the actual test data (not a re-weighted version, which we only used for training purposes)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from fairlearn.metrics import group_mean_prediction\n",
+    "\n",
+    "scores_pp = pd.Series(pp_estimator.predict(X_test, sensitive_features=A_test), name=\"scores_post\")\n",
+    "auc_pp = group_metric_as_df(\n",
+    "    \"auc_post\",\n",
+    "    group_roc_auc_score(y_test, scores_pp, A_test))\n",
+    "sel_pp = group_metric_as_df(\n",
+    "    \"selection_post\",\n",
+    "    group_mean_prediction(y_test, scores_pp, A_test))\n",
+    "display(pd.concat([auc_pp, sel_pp, auc_unmitigated], axis=1))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The original unmitigated scores have the demographic disparity around 0.6. We see that `ThresholdOptimizer` dramatically reduces the disparity to around 0.1 (the difference between selection rates). At the same time, the AUC in each subpopulation stays at around 0.65, a moderate drop from the original 0.72 and 0.74. This is a more favorable trade-off than the one achieved by model 35 above (with the disparity of 0.4 and the worst-case AUC of around 0.62). However, note that `ThresholdOptimizer` is a classifier, and so it can only work as a crude ranker. Additionally, `ThresholdOptimizer` uses the sensitive feature (in this instance race) at the prediction time, by applying a different threshold to unmitigated scores depending on race. In some use cases, these two properties might be undesirable. We next show how to obtain a classifier that also seeks to achieve low demographic disparity, but without requiring access to the sensitive feature at the evaluation time.\n",
```

Comment:
I think this paragraph is confusing, with some numbers appearing without explanation.

I believe that the figure of 0.6 for demographic disparity comes from the CDFs above? However, the disparity for the `ThresholdOptimizer` is said to just be the difference in the `selection_post` entries for black and white? If run `group_mean_prediction` on `scores_unmitigated` from the original unmitigated estimator above, then the numbers are 0.84 and 0.96, so a disparity of 0.12. Even using the 'raw' pass rates of 0.78 and 0.96, you only have a disparity of 0.18.

Reply:
what i'm doing is the exact opposite. run group_auc_roc on scores_pp.

Reply:
sorry.. not group_auc_roc, but the CDF based disparity on scores_pp.

Reply:
`plot_and_compare_cdfs(scores_pp, A_test)` is giving a rather ugly CDF, but does show a max disparity (actually, uniform disparity) of about 0.1. However, if that's the right way to measure the disparity (which is then consistent with what is done above), then shouldn't that be used here?

Reply:
I have added some clarification around where the numbers come from.

When it comes to demographic disparity--I think that a part of your confusion is driven by the fact that you know what the notebook looked like before (when it was showing the disparity in mean predictions for both classification and probabilistic classification). In the new version, we don't ever mention anything like average predictions, so perhaps this is not going to be too confusing. Note that the difference in selection rates is introduced as an equivalent way of obtaining demographic parity (which is defined as a difference between CDFs).

Also, this issue will get solved more transparently, once we have a dedicated function to evaluate demographic disparity.